src/Grpc/JsonTranscoding/src/Microsoft.AspNetCore.Grpc.JsonTranscoding/Internal/Json/MessageTypeInfoResolver.cs

```diff
@@ -0,0 +1,165 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections;
+using System.Diagnostics.CodeAnalysis;
+using System.Text.Json;
+using System.Text.Json.Serialization.Metadata;
+using Google.Protobuf;
+using Google.Protobuf.Reflection;
+using Google.Protobuf.WellKnownTypes;
+using Grpc.Shared;
+using Type = System.Type;
+
+namespace Microsoft.AspNetCore.Grpc.JsonTranscoding.Internal.Json;
+
+internal sealed class MessageTypeInfoResolver : DefaultJsonTypeInfoResolver
+{
+    private readonly JsonContext _context;
+
+    public MessageTypeInfoResolver(JsonContext context)
+    {
+        _context = context;
+    }
+
+    public override JsonTypeInfo GetTypeInfo(Type type, JsonSerializerOptions options)
+    {
+        var typeInfo = base.GetTypeInfo(type, options);
+
+        if (IsStandardMessage(type, out var messageDescriptor))
+        {
+            typeInfo.Properties.Clear();
```

Comment:
Not super important, but you might want to validate that `type` is tied to the internal object converter -- otherwise attempting to manipulate property metadata will result in an `InvalidOperationException`.
```suggestion
            Debug.Assert(typeInfo.TypeInfoKind == JsonTypeInfoKind.Object);

            typeInfo.Properties.Clear();
```

Reply:
IMessage will always be an object.

Something I did notice was types that had a converter specified on JsonSerializerOptions had a read-only Properties collection.

Reply:
also consider using JsonTypeInfo.CreateJsonTypeInfo here (you will need to assign CreateObject when you do that)

Reply:
> IMessage will always be an object.
> 
> Something I did notice was types that had a converter specified on JsonSerializerOptions had a read-only Properties collection.

Yeah, assuming somebody registers a custom converter for `IMessage` (i.e. the type is no longer associated with the converter that understands `JsonPropertyInfo` metadata) then `TypeInfoKind` would equal `JsonTypeInfoKind.None` and attempting to manipulate property metadata would result in an `InvalidOperationException`.